src/jvmMain/kotlin/io/ktor/io/JvmBuffer.kt

```diff
@@ -0,0 +1,156 @@
+     *
+     * The buffer is empty and prepared for write operations.
+     */
+    public constructor(capacity: Int) : this(
+        ByteBuffer.allocateDirect(capacity).limit(0),
+        ByteBufferPool.NoPool
+    )
+
+    /**
+     * Creates a new [JvmBuffer] instance with the [ByteBuffer] instance from the [pool].
+     *
+     * The buffer is empty and prepared for write operations.
+     */
+    public constructor(pool: ObjectPool<ByteBuffer> = ByteBufferPool.Default) : this(pool.borrow(), pool)
+
+    /**
+     * Provides access to the underlying [ByteBuffer].
+     *
+     * The [buffer.position] reflects [readIndex] and [buffer.limit] reflects [writeIndex].
+     *
+     * All modifications of the [ByteBuffer] is reflected by the [JvmBuffer] itself.
+     */
+    @Suppress("CanBePrimaryConstructorProperty")
+    public var buffer: ByteBuffer = buffer
+
+    override var readIndex: Int
+        get() = buffer.position()
+        set(value) {
+            buffer.position(value)
+        }
+
+    override var writeIndex: Int
+        get() = buffer.limit()
+        set(value) {
+            buffer.limit(value)
+        }
+
+    override val capacity: Int
+        get() = buffer.capacity()
+
+    override fun copyToArray(destination: ByteArray, startIndex: Int, endIndex: Int): Int {
+        require(startIndex >= 0) { "startIndex should be non-negative: $startIndex" }
+        require(startIndex <= endIndex) { "startIndex should be less than or equal to endIndex: $startIndex, $endIndex" }
+        require(endIndex <= destination.size) { "endIndex should be less than or equal to destination.size: $endIndex, ${destination.size}" }
+
+        val count = min(endIndex - startIndex, buffer.remaining())
+        buffer.get(destination, startIndex, count)
+        return count
+    }
+
+    /**
+     * Return the underlying buffer to the pool.
+     */
+    override fun close() {
+        pool.recycle(buffer)
+    }
+
+    override fun compact() {
+        buffer.compact()
+    }
+
+    override fun loadByteAt(index: Int): Byte = buffer.get(index)
+
+    override fun loadShortAt(index: Int): Short = buffer.getShort(index)
+
+    override fun loadIntAt(index: Int): Int = buffer.getInt(index)
+
+    override fun loadLongAt(index: Int): Long = buffer.getLong(index)
+
+    override fun storeByteAt(index: Int, value: Byte) {
+        randomAccess {
+            it.put(index, value)
+        }
+    }
+
+    override fun storeShortAt(index: Int, value: Short) {
+        randomAccess {
+            it.putShort(index, value)
+        }
+    }
+
+    override fun storeIntAt(index: Int, value: Int) {
+        randomAccess {
+            it.putInt(index, value)
+        }
+    }
+
+    override fun storeLongAt(index: Int, value: Long) {
+        randomAccess {
+            it.putLong(index, value)
+        }
+    }
+
+    override fun storeBufferAt(index: Int, value: Buffer): Int {
+        var current = index
```

Comment:
optimization if buffer is JvmBuffer?

Reply:
Lets postpone optimizations until we have a benchmark